### DISCUSSION TITLE
Support additional restrictions on typed parameters for methods

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#6817] [General]
I suggest supporting ReadOnlySpan<char> and Span<char> as parameters in typed `Dictionary<string, TValue>`, `HashSet<string>`, and so on classes.

Unfortunately, we can not use `HashSet<string>.Contains` method with `Span<char>` without creating a new string on the Heap or using the string pool.

```
public static class AClass
{
    private static HashSet<string> IgnoryWords = new HashSet<string>()
        {
            "the", "of", "in", "from", "by", "with", "and", "or", "for", "to", "at", "a"
        };

    // Using string pool any way will copy span memory to pooled string
    public static bool IsIgnoryWord(this ReadOnlySpan<char> word) =>
        IgnoryWords.Contains(new string(word)); // creates string on Heap 🤦‍♂️
}
```

We can use a string pool but it isn't easy and will copy span memory to pooled string in any way.

So I suggest enabling to define methods in classes and interfaces with additional restrictions on typed parameters.

For Example
```
public class HashSet<T>
{
	public bool Contains(ReadOnlySpan<char> span) **where T : string** // additional restrictions on typed parameter!!!
		=> throw new NotImplementedException();
}
```

Now that syntax is disabled, see https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0080

I suggest enabling that syntax to support additional methods when typed parameters have some additional restrictions only for that methods.

So `Contains(ReadOnlySpan<char> span)` method will have only string HashSet `HashSet<string>`.

**Comment by svick** [comment]
The request for a span-based indexing into a `Dictionary<string, TValue>` is tracked by https://github.com/dotnet/runtime/issues/27229 (which also briefly mentioned `HashSet<string>`). As mentioned there, it could be implemented today by the runtime using extension methods:

```c#
public static class DictionaryExtensions
{
    public static bool ContainsKey<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key);
    public static bool Remove<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key);
    public static bool TryGetValue<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key, out TValue value);
    public static T GetValue<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key);
}
```

And having a decent alternative makes the proposed feature less useful.

Another hurdle is that adding this feature would very likely require runtime changes.

**Reply by AlexRadch** [reply]
New C# syntax will look cleaner than extension methods for which the class should open internal methods, fields, and properties.

**Reply by CyrusNajmabadi** [reply]
Accessing internals is definitely not allowed. That would violate safety for tons of types.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/1315

**Comment by koszeggy** [comment]
Just a side note in case you can't wait for it to be added to .NET: my [core libraries](https://github.com/koszeggy/KGySoft.CoreLibraries#stringkeyeddictionarytvalue) contain a [`StringKeyedDictionary<TValue>`](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_Collections_StringKeyedDictionary_1.htm) class since .NET 5.0 (but older targets are also supported) that supports lookups with `ReadOnlySpan<char>` instances ([NuGet](https://www.nuget.org/packages/KGySoft.CoreLibraries), [source](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/master/KGySoft.CoreLibraries/Collections/StringKeyedDictionary.cs)).

**Comment by CyrusNajmabadi** [comment]
Moving to runtime as this appears to be an API request.

**Reply by AlexRadch** [reply]
I write the request to extend the CSharp language, **not** API! API is the only example of how **additional restrictions on typed parameters for methods** can be used.

Extension methods have similar possibilities without extending the CSharp language but have many disadvantages. For example, you should open class internal methods, fields, properties, and so on to implement extension methods.

**Reply by CyrusNajmabadi** [reply]
It's not clear why that can't be an extension method on. `HashSet<string>`.

**Reply by AlexRadch** [reply]
It can, but to implement such extension method, you should open class internal methods, fields, properties.

**Reply by AlexRadch** [reply]
Also, the extension method can not be virtual, and so on disadvantages.

**Reply by AlexRadch** [reply]
Can you return this discussion to the [C# Language repository](https://github.com/dotnet/csharplang/discussions/categories/language-ideas)?

**Reply by koszeggy** [reply]
If you only want to be able to use `Contains[Key]`/`TryGetValue`/etc. with `ReadOnlySpan<char>` for string keys, why is it so important for you to have a new language feature after all? You would not use the new C# syntax, just use a feature specific to `Dictionary<string, TValue>`/`HashSet<string>`.

> New C# syntax will look cleaner than extension methods for which the class should open internal methods, fields, and properties.

A public extension method in the core library can use the internals of `Dictionary<TKey, TValue>` just like as [`CollectionMarshal`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal) accesses the internals of `List<T>` and `Dictionary<TKey, TValue>` without exposing anything to other assemblies.

And an extension method specifically for string key is cleaner than polluting the generic class itself by exceptional logic specific for the various possible generic arguments.

But a specialized `StringKeyDictionary<TValue>` could maybe even cleaner. [#27229](https://github.com/dotnet/runtime/issues/27229) plans it to be derived from `Dictionary<string, TValue>`, whereas [my already existing version](https://github.com/koszeggy/KGySoft.CoreLibraries/blob/master/KGySoft.CoreLibraries/Collections/StringKeyedDictionary.cs) is a complete implementation from scratch so it can use a special [comparer](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_CoreLibraries_StringSegmentComparer.htm) that is able to compare a string to `ReadOnlySpan<char>` without extra allocations.

**Reply by AlexRadch** [reply]
I think the new syntax would be useful not only for Dictionary<string, TValue>/HashSet<string> but for many other methods where the method must have additional restrictions for typed parameters. Especially if the methods must be virtual because virtual methods cannot be implemented as extension methods.

My suggestion is not to add Span support to Dictionary<string, TValue>/HashSet<string> but for all similar cases also. I published this suggestion to [C# Language repository](https://github.com/dotnet/csharplang/discussions/categories/language-ideas), but @CyrusNajmabadi moved it to this repository and I can not move it back.

**Reply by koszeggy** [reply]
> @CyrusNajmabadi moved it to this repository and I can not move it back

This is the csharplang repository. He moved it back about 11 hours ago. Isn't this what you requested?

> I think the new syntax would be useful not only for Dictionary<string, TValue>/HashSet but for many other methods where the method must have additional restrictions for typed parameters.

Could you provide an example?

> Especially if the methods must be virtual because virtual methods cannot be implemented as extension methods.

This is the 2nd time you use this reasoning but an example would be useful because it's a bit unclear what you want to achieve. You can "override" an extension method behavior by specifying a new extension method with the same name on a more specialized type. A nice example is LINQ where the "base" operations are in the `Enumerable` class, just like [`Contains(IEnumerable<T>, TSource)`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.contains), which is "overridden" in [`Contains(IQueryable<T>, TSource)`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.queryable.contains), [`Contains(ParallelQuery<TSource>, TSource)`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.parallelenumerable.contains), etc.
